torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2936,12 +2930,17 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # reduce_dtype matches the param dtype.
                     param.grad.data = param.grad.data.to(self.mixed_precision.reduce_dtype)
 
-                if self.gradient_predivide_factor > 1 and self.communication_hook is None:
-                    # Average grad by pre-division factor. Together pre- and post-division factors
-                    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
-                    # This is a two-step process to avoid potential underflow and overflow.
-                    param.grad.div_(self.gradient_predivide_factor)
-
+                if self._exec_order_data.is_first_iter:
+                    # For all sharding strategies communication is performed through `communication_hook`:
```

Comment:
Same as above. How can these attributes be `None`?

In general, I am not supportive of these kinds of inline comments. If you find yourself duplicating comments in multiple places, then it is a sign that that code should be refactored.

Reply:
@awgu This can be manually set to `None` by user at some point, e.g. `model.comm_hook=None`. I had a request from previous PR to add the check, that `comm_hook` is not `None` before calling it

Reply:
Can you link the comment so I can get the context? Maybe we should make the `comm_hook` private (i.e. `_comm_hook`) and only have a getter and not a setter?

Reply:
@awgu https://github.com/pytorch/pytorch/pull/79833#discussion_r910616524

Reply:
@awgu I understand why we may need to keep `comm_hook` private, but could you clarify why to get rid of the setter? i.e. how `register_comm_hook` will set `comm_hook` then? I am probably misunderstanding something. though

Reply:
Discussed offline.